Non-primary Tap Stream source names are changed after deploy, StackOverflow Exception is thrown when inspecting on dashboard

I initially created some tap streams as following:

Stream "base":
`workflow-service --server.port=8085 | log`   
Stream "comparator"
`:base.workflow-service > comparator-service | rabbit --routing-key=comparator --username=guest --password=******`
Stream "mapper"
`:base.workflow-service > mapper-one | rabbit --routing-key=mapper --username=guest --password=******            `

After few deploy and undeploy, non-primary tab streams are changed as following:
Stream "comparator"
`:comparator > comparator-service | rabbit --spring.rabbitmq.password=guest --rabbit.routing-key=comparator --spring.rabbitmq.username=guest  `
Stream "mapper" 
`:mapper > mapper-one | rabbit --spring.rabbitmq.password=guest --rabbit.routing-key=mapper --spring.rabbitmq.username=guest`


When inspecting the 'base' stream on the dashboard, it originally shows all the forks and tap streams. After changed, StackOverFlow exception was thrown:

> 2018-04-18 15:22:07.744 ERROR 32747 --- [nio-9393-exec-3] o.s.c.d.s.c.RestControllerAdvice         : Caught exception while handling a request
> 
> org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.StackOverflowError
> 	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:982) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970) [spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861) [spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:635) [tomcat-embed-core-8.5.27.jar!/:8.5.27]
> 
> Caused by: java.lang.StackOverflowError: null
> 	at java.util.Collections$UnmodifiableCollection.size(Collections.java:1030) ~[na:1.8.0_162]
> 	at org.springframework.cloud.dataflow.core.dsl.Tokens.hasNext(Tokens.java:131) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.Tokens.next(Tokens.java:199) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatDestinationReference(StreamParser.java:302) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatSourceDestination(StreamParser.java:257) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatStream(StreamParser.java:137) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.parse(StreamParser.java:74) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.server.controller.StreamDefinitionController.findRelatedDefinitions(StreamDefinitionController.java:296) ~[spring-cloud-dataflow-server-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.server.controller.StreamDefinitionController.findRelatedDefinitions(StreamDefinitionController.java:302) ~[spring-cloud-dataflow-server-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.server.controller.StreamDefinitionController.findRelatedDefinitions(StreamDefinitionController.java:302) ~[spring-cloud-dataflow-server-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]

--- a/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDefinitionController.java
+++ b/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDefinitionController.java
@@ -68,6 +68,7 @@ import org.springframework.web.bind.annotation.RestController;
  * @author Oleg Zhurakousky
  * @author Glenn Renfro
  * @author Christian Tzolov
+ * @author Andy Clement
  */
 @RestController
 @RequestMapping("/streams/definitions")
@@ -250,8 +251,8 @@ public class StreamDefinitionController {
 			if (sn.getSourceDestinationNode() != null) {
 				String nameComponent = sn.getSourceDestinationNode().getDestinationName();
 				if (nameComponent.equals(currentStreamName) || nameComponent.startsWith(indexedStreamName)) {
-					relatedDefinitions.add(definition);
-					if (nested) {
+					boolean isNewEntry = relatedDefinitions.add(definition);
+					if (nested && isNewEntry) {
 						findRelatedDefinitions(definition, definitions, relatedDefinitions, true);
 					}
 				}

--- a/spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java
+++ b/spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java
@@ -83,6 +83,7 @@ import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.
  * @author Janne Valkealahti
  * @author Gunnar Hillert
  * @author Glenn Renfro
+ * @author Andy Clement
  */
 @RunWith(SpringRunner.class)
 @SpringBootTest(classes = TestDependencies.class)
@@ -181,6 +182,39 @@ public class StreamControllerTests {
 		assertTrue(response.contains(":myStream1.time > log"));
 		assertTrue(response.contains("time | log"));
 		assertTrue(response.contains("\"totalElements\":3"));
+	}
+
+	@Test
+	public void testFindRelatedStreams_gh2150() throws Exception {
+		assertEquals(0, repository.count());
+		// Bad definition, recursive reference
+		mockMvc.perform(post("/streams/definitions/").param("name", "mapper")
+				.param("definition", ":mapper.time > log")
+				.accept(MediaType.APPLICATION_JSON)).andDo(print()).andExpect(status().isCreated());
+		assertEquals(1, repository.count());
+		String response = mockMvc
+				.perform(get("/streams/definitions/mapper/related").param("nested", "true").accept(MediaType.APPLICATION_JSON)).andReturn()
+				.getResponse().getContentAsString();
+		assertTrue(response.contains(":mapper.time > log"));
+		assertTrue(response.contains("\"totalElements\":1"));
+	}
+
+	@Test
+	public void testFindRelatedStreams2_gh2150() throws Exception {
+		// bad streams, recursively referencing via each other
+		mockMvc.perform(post("/streams/definitions/").param("name", "foo")
+				.param("definition", ":bar.time > log")
+				.accept(MediaType.APPLICATION_JSON)).andDo(print()).andExpect(status().isCreated());
+		mockMvc.perform(post("/streams/definitions/").param("name", "bar")
+				.param("definition", ":foo.time > log")
+				.accept(MediaType.APPLICATION_JSON)).andDo(print()).andExpect(status().isCreated());
+		assertEquals(2, repository.count());
+		String response = mockMvc
+				.perform(get("/streams/definitions/foo/related").param("nested", "true").accept(MediaType.APPLICATION_JSON)).andReturn()
+				.getResponse().getContentAsString();
+		assertTrue(response.contains(":foo.time > log"));
+		assertTrue(response.contains(":bar.time > log"));
+		assertTrue(response.contains("\"totalElements\":2"));
 	}
 
 	@Test

